Userland/Libraries/LibVirtGPU/CommandBufferBuilder.cpp

```diff
@@ -0,0 +1,310 @@
+ */
+
+#include <AK/StringView.h>
+#include <Kernel/API/VirGL.h>
+#include <sys/ioctl_numbers.h>
+
+#include <LibVirtGPU/CommandBufferBuilder.h>
+#include <LibVirtGPU/VirGLProtocol.h>
+
+namespace VirtGPU {
+
+static u32 encode_command(u32 length, u32 mid, Protocol::VirGLCommand command)
+{
+    u32 command_value = to_underlying(command);
+    return (length << 16) | ((mid & 0xff) << 8) | (command_value & 0xff);
+};
+
+class CommandBuilder {
+public:
+    CommandBuilder() = delete;
+    CommandBuilder(Vector<u32>& buffer, Protocol::VirGLCommand command, u32 mid)
+        : m_buffer(buffer)
+        , m_start_offset(buffer.size())
+        , m_command(command)
+        , m_command_mid(mid)
+    {
+        m_buffer.append(0);
+    }
+    void appendu32(u32 value)
+    {
+        VERIFY(!m_finalized);
+        m_buffer.append(value);
+    }
+    void appendf32(float value)
+    {
+        VERIFY(!m_finalized);
+        m_buffer.append(bit_cast<u32>(value));
+    }
+    void appendf64(double value)
+    {
+        VERIFY(!m_finalized);
+        m_buffer.append(0);
+        m_buffer.append(0);
+        auto* depth = (u64*)(&m_buffer[m_buffer.size() - 2]);
+        *depth = bit_cast<u64>(value);
+    }
+    void append_string_null_padded(StringView string)
+    {
+        VERIFY(!m_finalized);
+        // Remember to have at least one null terminator byte
+        auto length = string.length() + 1;
+        auto num_required_words = (length + sizeof(u32) - 1) / sizeof(u32);
+        m_buffer.resize(m_buffer.size() + num_required_words);
+        char* dest = (char*)&m_buffer[m_buffer.size() - num_required_words];
+        memcpy(dest, string.characters_without_null_termination(), string.length());
+        // Pad end with null bytes
+        memset(&dest[string.length()], 0, 4 * num_required_words - string.length());
+    }
+    void finalize()
+    {
+        if (!m_finalized) {
+            m_finalized = true;
+            size_t num_elems = m_buffer.size() - m_start_offset - 1;
+            m_buffer[m_start_offset] = encode_command(num_elems, m_command_mid, m_command);
+        }
+    }
+    ~CommandBuilder()
+    {
+        if (!m_finalized)
+            finalize();
+    }
+
+private:
+    Vector<u32>& m_buffer;
+    size_t m_start_offset;
+    Protocol::VirGLCommand m_command;
+    u32 m_command_mid;
+    bool m_finalized { false };
+};
+
+void CommandBufferBuilder::append_set_tweaks(u32 id, u32 value)
+{
+    CommandBuilder builder(m_buffer, Protocol::VirGLCommand::SET_TWEAKS, 0);
+    builder.appendu32(id);
+    builder.appendu32(value);
+}
+
+void CommandBufferBuilder::append_transfer3d(Protocol::ResourceID resource, size_t width, size_t height, size_t depth, size_t direction)
+{
+    CommandBuilder builder(m_buffer, Protocol::VirGLCommand::TRANSFER3D, 0);
+    builder.appendu32(resource.value()); // res_handle
+    builder.appendu32(0);                // level
+    builder.appendu32(242);              // usage
```

Comment:
Can you make this `242` somewhat less magic?

Reply:
I legitimately have no idea why the usage has to be 242. What I do know however is that the demo sometimes failed if it wasn't 242.this could probably be factored out as more of the protocol gets deciphered

Reply:
I've quickly checked but can't find anything on this either. Maybe dedicate a FIXME to this value? It stands out like a sore thumb